modules/renter/worker.go

```diff
@@ -0,0 +1,287 @@
+		pieceIndex uint64
+
+		chunkDownload *chunkDownload
+
+		// resultChan is a channel that the worker will use to return the
+		// results of the download.
+		resultChan chan finishedDownload
+	}
+
+	// finishedDownload contains the data and error from performing a download.
+	finishedDownload struct {
+		chunkDownload *chunkDownload
+		data          []byte
+		err           error
+		pieceIndex    uint64
+		workerID      types.FileContractID
+	}
+
+	// finishedUpload contains the Merkle root and error from performing an
+	// upload.
+	finishedUpload struct {
+		chunkID    chunkID
+		dataRoot   crypto.Hash
+		err        error
+		pieceIndex uint64
+		workerID   types.FileContractID
+	}
+
+	// uploadWork contains instructions to upload a piece to a host, and a
+	// channel for returning the results.
+	uploadWork struct {
+		// data is the payload of the upload.
+		chunkID    chunkID
+		data       []byte
+		file       *file
+		pieceIndex uint64
+
+		// resultChan is a channel that the worker will use to return the
+		// results of the upload.
+		resultChan chan finishedUpload
+	}
+
+	// A worker listens for work on a certain host.
+	worker struct {
+		// contractID specifies which contract the worker specifically works
+		// with.
+		contractID types.FileContractID
+
+		// If there is work on all three channels, the worker will first do all
+		// of the work in the download chan, then all of the work in the
+		// priority upload chan, and finally all of the work in the upload
+		// chan.
+		//
+		// A busy higher priority channel is able to entriely starve all of the
+		// channels with lower priority.
+		downloadChan         chan downloadWork // higher priority than all uploads
+		killChan             chan struct{}     // highest priority
+		priorityDownloadChan chan downloadWork // higher priority than downloads (used for user-initiated downloads)
+		uploadChan           chan uploadWork   // lowest priority
+
+		// recentUploadFailure documents the most recent time that an upload
+		// has failed. The repair loop ignores workers that have had an upload
+		// failure in the past two hours.
+		recentUploadFailure time.Time
+
+		// Utilities
+		renter *Renter
+	}
+)
+
+// download will perform some download work.
+func (w *worker) download(dw downloadWork) {
+	d, err := w.renter.hostContractor.Downloader(w.contractID)
+	if err != nil {
+		dw.resultChan <- finishedDownload{dw.chunkDownload, nil, err, dw.pieceIndex, w.contractID}
+		return
+	}
+	defer d.Close()
+
+	data, err := d.Sector(dw.dataRoot)
+	dw.resultChan <- finishedDownload{dw.chunkDownload, data, err, dw.pieceIndex, w.contractID}
+}
+
+// upload will perform some upload work.
+func (w *worker) upload(uw uploadWork) {
+	e, err := w.renter.hostContractor.Editor(w.contractID)
+	if err != nil {
+		w.recentUploadFailure = time.Now()
+		select {
+		case uw.resultChan <- finishedUpload{uw.chunkID, crypto.Hash{}, err, uw.pieceIndex, w.contractID}:
+		case <-w.renter.tg.StopChan():
+		}
```

Comment:
may want to define a helper for this. You can do it inline with `w.upload` so that it captures its variables:
```go
func (w *worker) upload(uw uploadWork) {
    sendResult := func(u finishedUpload) {
        select {
        case uw.resultChan <- u:
        case <-w.renter.tg.StopChan():
        }
    }
    // ...
```
Also, I noticed that `w.download` doesn't have an equivalent for stopping early.